File: apps/api/src/monitoring/ingest/ingest-token.service.ts
import { Injectable } from '@nestjs/common';
import { randomBytes, createHash } from 'node:crypto';
import { PrismaService } from '../../prisma/prisma.service';

const hash = (token: string) => createHash('sha256').update(token).digest('hex');

/**
 * Per-org ingest secret: the plaintext is shown once at (re)generation; only its
 * sha256 is stored. verify() resolves a presented token back to its org (or null).
 */
@Injectable()
export class IngestTokenService {
  constructor(private readonly prisma: PrismaService) {}

  async createOrRotate(organizationId: string): Promise<string> {
    const secret = randomBytes(32).toString('base64url');
    const tokenHash = hash(secret);
    await this.prisma.monitoringIngestToken.upsert({
      where: { organizationId },
      create: { organizationId, tokenHash },
      update: { tokenHash },
    });
    return secret; // shown once; only the hash is persisted
  }

  async verify(token: string): Promise<string | null> {
    if (!token) return null;
    const row = await this.prisma.monitoringIngestToken.findFirst({
      where: { tokenHash: hash(token) },
      select: { organizationId: true },
    });
    return row?.organizationId ?? null;
  }
}

File: src/Modules/Monitoring/NodeScope.Modules.Monitoring.Application/Ingest/IngestTokenService.cs
using System.Buffers.Text;
using System.Security.Cryptography;
using NodeScope.Modules.Monitoring.Application.Agents;

namespace NodeScope.Modules.Monitoring.Application.Ingest;

/// <summary>Persistence for the one-per-org ingest token (hash only).</summary>
public interface IIngestTokenRepository
{
    /// <summary>Upserts the org's token hash (rotation overwrites; the old token dies instantly).</summary>
    public Task UpsertAsync(string organizationId, string tokenHash, CancellationToken cancellationToken);

    /// <summary>The owning org of a token hash, or null.</summary>
    public Task<string?> FindOrganizationByTokenHashAsync(string tokenHash, CancellationToken cancellationToken);
}

/// <summary>
/// The per-org ingest secret (Node's <c>IngestTokenService</c>): plaintext shown once at
/// (re)generation, only its SHA-256 stored; rotation is the only revocation.
/// </summary>
public sealed class IngestTokenService
{
    private readonly IIngestTokenRepository _repo;

    public IngestTokenService(IIngestTokenRepository repo)
    {
        _repo = repo;
    }

    public async Task<string> CreateOrRotateAsync(string organizationId, CancellationToken cancellationToken)
    {
        var secret = Base64Url.EncodeToString(RandomNumberGenerator.GetBytes(32));
        await _repo.UpsertAsync(organizationId, AgentTokenService.Sha256Hex(secret), cancellationToken);
        return secret;
    }

    /// <summary>The org a presented token belongs to, or null.</summary>
    public Task<string?> VerifyAsync(string token, CancellationToken cancellationToken) =>
        string.IsNullOrEmpty(token)
            ? Task.FromResult<string?>(null)
            : _repo.FindOrganizationByTokenHashAsync(AgentTokenService.Sha256Hex(token), cancellationToken);
}
